File: algorithm/leetcode/077_combinations/combinations.py
from copy import copy

# my solution
class Solution:
    def combine(self, n, k):
        res = []
        for i in range(1, n-k+2):
            self.backtrace(i, n, k, [], res)
        return res

    def backtrace(self, cur, n, k, tmp_comb, res):
        tmp_comb.append(cur)
        if len(tmp_comb) == k:
            res.append(tmp_comb)
            return
        if cur > n:
            return
        for i in range(cur+1, n-k+2+len(tmp_comb)):
            self.backtrace(i, n, k, copy(tmp_comb), res)


# clean solution
class Solution1:
    def combine(self, n, k, start=1)
    if k == 0:
        return [[]]
    return [
        [i, *j]
        for i in range(start, n+1) 
        for j in self.combine(n, k-1, i+1)]


# build-in pkg solution
class Solution1:
    def combine(self, n, k, start=1)
        from itertools import combinations
        return combinations(range(1, n+1), k)


if __name__ == '__main__':
    s = Solution()
    print(s.combine(1, 1))

File: algorithm/leetcode/077_combinations/combinations.go
package leetcode077

func combine(n int, k int) [][]int {
	res := make([][]int, 0)

	var backtrack func(int, []int)
	backtrack = func(cur int, tmpList []int) {
		if len(tmpList) == k {
			t := make([]int, len(tmpList))
			copy(t, tmpList)
			res = append(res, t)
			return
		}
		for i := cur; i <= n; i++ {
			tmpList = append(tmpList, i)
			backtrack(i+1, tmpList)
			tmpList = tmpList[:len(tmpList)-1]
		}
	}

	return res
}
